### DISCUSSION TITLE
is possible to use mermaid beta diagrams (architecture-beta)?

### DISCUSSION THREAD

**Discussion by Standard-IO** [facebook/docusaurus#10508] [Q&A] [answered]
I'm trying to use mermaid Architecture Diagrams (architecture-beta) but I just got the 
> UnknownDiagramError: No diagram type detected matching given configuration for text: architecture-beta

is possible to use  it?

I'm currently using  the next versions
```
    "@docusaurus/core": "3.5.2"
    "@docusaurus/preset-classic": "3.5.2",
    "@docusaurus/theme-mermaid": "^3.5.2",
```

**Comment by slorber** [comment, marked as answer]
This feature is Mermaid 11.1+ (https://mermaid.js.org/syntax/architecture)

Our dependency is `"mermaid": "^10.4.0",`, so no you can't use Mermaid 11.1+ features yet.

However the list of v11 breaking changes is small:
https://github.com/mermaid-js/mermaid/releases/tag/v11.0.0

It is likely that we could add support for v11 simply by updating our semver range and allowing to use v11 versions. It probably doesn't require a Docusaurus breaking change (ie waiting for Docusaurus v4). Will see if I can get it working.

**Reply by slorber** [reply]
Seems to work fine:

![CleanShot 2024-09-19 at 12 43 54](https://github.com/user-attachments/assets/2c544b2c-ffd8-4936-9f01-d86e46f18c37)


PR: https://github.com/facebook/docusaurus/pull/10510

**Reply by Standard-IO** [reply]
That's great.

**Reply by evolutioned** [reply]
This would be useful. Could you provide some guidance on how to enable this in our projects?

**Reply by slorber** [reply]
If you want to use V11 you have to force the resolution to v11 using your package manager features (for example Yarn resolutions)

Otherwise you'll have to wait for the Docusaurus v3.6 release coming in the next few weeks.

**Reply by evolutioned** [reply]
Thank you.  We can wait and upgrade to 3.6.

**Comment by RRQM** [comment]
Please support Mermaid 11.1 as soon as possible, I really need some of the features it includes. thank you

**Reply by slorber** [reply]
https://docusaurus.io/blog/releases/3.6#mermaid

**Reply by RRQM** [reply]
Thank you for your reply. I have completed the upgrade and wish Docusaurus continued success.

**Comment by wgordon17** [comment]
It definitely seems like I'm able to utilize the new architecture diagrams, however I can't figure out how to register additional icons (https://mermaid.js.org/config/icons.html). Has anyone had luck with this?

**Reply by slorber** [reply]
We don't provide an "official way" of doing that, but calling `mermaid.registerIconPacks` in a Docusaurus client module should work. Alternatively, you can try to swizzle the `@theme/Mermaid` component and put that here.

**Reply by wgordon17** [reply]
I'm good with no official way for that 👍 Tried using a client module, didn't seem to work so well. I'll give swizzling a try, thanks!

**Reply by slorber** [reply]
What didn't work exactly

Show me your attempt using docusaurus.new and I'll tell you what you did wrong. I can't help without an error message or something more concrete.

**Reply by wgordon17** [reply]
(It took me _wayyy_ to long to figure out why stackblitz.com wouldn't work for me! Had to disable uBlock...anyway...)

Here's my project https://stackblitz.com/edit/github-havqfqht?file=src%2Fscripts%2Fmermaid_icons.js,docs%2Fintro.md,docusaurus.config.js

It's not an error message that I can _see_, it's just that instead of the _intended_ logo, it loads as a `?` (which means that the logo pack just isn't getting registered).

I appreciate the help!

**Reply by wgordon17** [reply]
Oh, well f-me! Why is it working now?!

Ok, for clarity...and anyone else that happens to stumble on this later...

`docusaurus.config.js`
```javascript
...
  themes: ['@docusaurus/theme-mermaid'],
  markdown: {
    mermaid: true,
  },
  clientModules: [require.resolve('./src/scripts/mermaid_icons.js')],
...
```

`src/scripts/mermaid_icons.js`
```javascript
import mermaid from 'mermaid';
import { icons } from '@iconify-json/logos';
mermaid.registerIconPacks([
  {
    name: icons.prefix, // To use the prefix defined in the icon pack
    icons,
  },
]);
```

And then this works!
```
```mermaid
architecture-beta
    group api(logos:aws)[API]

    service db(logos:aws-rds)[Database] in api
    service disk1(disk)[Storage] in api
    service disk2(disk)[Storage] in api
    service server(logos:aws-ec2)[Server] in api

    db:L -- R:server
    disk1:T -- B:server
    disk2:T -- B:db
```

![image](https://github.com/user-attachments/assets/acdade89-ca4c-44ed-affd-11f227f79f0a)

Thanks all the same @slorber!

**Reply by jovezhong** [reply]
I cannot load extra icons, just got a big ? in the box...

**Reply by wgordon17** [reply]
@jovezhong Have you

- [ ] Installed the `@iconify-json/logos` logo pack **and** the `@docusaurus/theme-mermaid` theme into `package.json`?
- [ ] Enabled mermaid in `docusaurus.config.js`?
- [ ] Enabled the mermaid _theme_ in `docusaurus.config.js`?
- [ ] Created a `mermaid_icons.js` script to register the logo pack?
- [ ] Added the `mermaid_icons.js` script to `clientModules` in `docusaurus.config.js`?
- [ ] Referenced the logo pack correctly in your mermaid block by prefixing `logos:`?
- [ ] Ensured that the logo exists in https://icon-sets.iconify.design/logos/?
- [ ] Restarted `docusaurus`, assuming you're testing on the local dev server?

**Reply by jovezhong** [reply]
Thank you, @wgordon17 for the checklist. I did do all of them, but still cannot load a custom icon such as aws-ec2. I was on a private brunch. Let me make it public. 1 sec

**Reply by jovezhong** [reply]
here is the commit https://github.com/timeplus-io/docs/pull/288/commits/01b8faaa7a852141525a6bf0ae40adb5e58a8002
and the preview site https://67c7215ced80cd000884eeac--elastic-elion-a958b6.netlify.app/tired-storage I changed the icon on the right to be aws-s3 and got a big question mark. If you can help to check what I did wrong, that will be great. Thanks.

**Reply by wgordon17** [reply]
Sorry, totally unrelated, do you mean "Tiered storage" instead of "Tired storage"?

**Reply by wgordon17** [reply]
Alrighty, you're not gonna like this...but I can't reproduce the issue 🙈 I have it working correctly on a docusaurus.new test https://stackblitz.com/edit/github-yydbha4z?file=docusaurus.config.js,src%2Fscripts%2Fmermaid_icons.js,package.json,docs%2Fintro.md

**Reply by jovezhong** [reply]
Sorry, I will change to "Tiered", not a native speaker and drafted that page in the midnight.. any clue why the icon cannot be loaded. I guess I am using the latest version.

**Reply by wgordon17** [reply]
Have you tried it locally, _not_ based on the preview?? I just cloned your repo, and checked out the `release/2.8` branch. Believe it or not, with an `npm start`, it's showing the logo just fine for me ¯\\\_(ツ)_/¯ 

![image](https://github.com/user-attachments/assets/1aefc859-3545-4342-8240-27a2e72a1700)

I'm _guessing_ something might be wrong/stale in your preview system possibly??

**Reply by wgordon17** [reply]
Yup, even a local `npm run build` and testing the static HTML brings up the working logos just fine and without any code changes. I'm willing to somehow blame netlify 😉

**Reply by jovezhong** [reply]
usually I ran `yarn run start` but I also tried `npm start`. Same issue locally and on the preview site. Again, thanks for looking into this. I don't see related error messages in the JS console...  https://deploy-preview-288--elastic-elion-a958b6.netlify.app/tiered-storage is the new link
